src/languageclient.ts

```diff
@@ -0,0 +1,266 @@
+                            fileEvents: workspace.createFileSystemWatcher('**/*.tf'),
+                        }
+                    };
+
+                    // Create the language client and start the client.
+                    let langClient = new LanguageClient(
+                        'terraformLanguageServer',
+                        'Language Server',
+                        serverOptions,
+                        clientOptions,
+                        true
+                    );
+
+                    ExperimentalLanguageClient.client = langClient;
+
+                    langClient.trace = 2;
+
+                    langClient.onReady().then(() => {
+                        const capabilities = langClient.initializeResult && langClient.initializeResult.capabilities;
+                        if (!capabilities) {
+                            return vscode.window.showErrorMessage('The language server is not able to serve any features at the moment.');
+                        } else {
+
+                            progress.report({
+                                message: "Language Server Ready",
+                                increment: 100
+                            })
+                            ExperimentalLanguageClient.isRunning = true;
+
+                            resolve();
+                        }
+                    });
+
+                    progress.report({
+                        message: "Initializing Language Server",
+                        increment: 70
+                    })
+
+                    // Start the client. This will also launch the server
+                    this.ctx.subscriptions.push(langClient.start());
+
+                    try {
+                        // After server is running lets do a check to see if a newer one exists
+                        const octokit = new Octokit();
+                        const release = await octokit.repos.getLatestRelease({
+                            owner: 'juliosueiras',
+                            repo: 'terraform-lsp'
+                        });
+
+                        if (this.currentReleaseId !== release.data.id.toString()) {
+                            vscode.window.showInformationMessage(`A newer Lanaguage Server is available ${release.data.tag_name}. Run the 'Install/Update Language Server' command to update`);
+                        }
+                    } catch (e) {
+                        return vscode.window.showErrorMessage(`Failed checking for newer language server version. Error: ${e}`);
+                    }
+
+                } catch (e) {
+                    vscode.window.showErrorMessage(`Failed starting language server. Error: ${e}`);
+                    reject(e);
+                }
+            });
+        });
+
+    }
+
+    private async installCommonProviders(serverLocation: string) {
+        let terminal = vscode.window.createTerminal("Terraform: Install providers");
```

Comment:
do you think it makes sense to create a `Terraform` terminal and reuse all over the place?

Reply:
I like the idea, the problem would be that I wait on the process exiting to know the install of the providers is complete so I kinda need it to be single use.